### ISSUE TITLE
Add a way to opt out of type hints on specific types

### ISSUE THREAD

**Issue by zesterer** [rust-analyzer/rust-analyzer#11240]
Hello!

I'm the author of [`chumsky`](https://github.com/zesterer/chumsky/), a parser combinator crate. As I'm sure you know, parser combinators involve combining simple parsers together with the type system to produce parsers for more complex syntaxes.

Unfortunately, this can result in very large types that, although very useful to the compiler, are almost entirely useless for the programmer trying to work with them (each parser has a virtually identical API to any other, and so the type conveys no real meaning: only the traits that it implements matter).

I'm not personally a user of `rust-analyzer`, but I have a few users that are, and I've noticed that they're getting rather intrusive type annotations within combinator code, like so:

![image](https://user-images.githubusercontent.com/3268430/148662232-7cc0c501-4917-411d-9f03-bce33ab8c50c.png)

It would be great is there were a way to opt out of inline type hints (perhaps with an override) that could be controlled by crate developers. Perhaps it would be possible to have some sort of `#[rust_analyzer(no_hint)]` attribute that I can add to my combinator types that would be understood by `rust-analyzer`?

**Comment by jhgg** [comment]
I don't think we should allow crate developers to turn off inlay hints - I fear this might lead to even more confusion for the end user. But perhaps we should endeavor to render something more sensible for inlay hints in places like this, similar to what is done for iterative chains already.

**Comment by 1Dragoon** [comment]
Couldn't type aliases solve this?

**Comment by flodiebold** [comment]
No, because 1. rust-analyzer doesn't use type aliases when displaying types, and doing so is kind of complicated, and 2. type aliases can't really simplify such a nested type much anyway. `padded()` has to return `Padded<Self>`, so a type alias wouldn't help. Returning `impl Parser` would help, but has other problems.

**Comment by Veykril** [comment]
Ideally via https://github.com/rust-lang/rust-analyzer/issues/11556 would be my first thought
